### DISCUSSION TITLE
Body parts switching

### DISCUSSION THREAD

**Discussion by auesro** [talmolab/sleap#1000] [Help!] [answered]
Hi everyone,

Is there any option to correct node's identity switching? We have created a model trained with more than 1000 frames but we still get some fast body parts identity switch. Can we correct this in the analysis phase?

Here is a visual example of what I refer to (check the fast switch between the grey/black/pink nodes):
https://user-images.githubusercontent.com/24290290/196726770-7775a80b-a2d3-4e15-8537-4fa671156b4b.mp4

Thanks!

A

**Comment by roomrys** [comment, marked as answer]
Hi @auesro,

Can you try adjusting the [`--peak_threshold`](https://sleap.ai/develop/guides/cli.html#sleap-track:~:text=up%20(PAF)%20models.%0A%20%20%2D%2D-,peak_threshold,-PEAK_THRESHOLD%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20Minimum%20confidence) value when running inference through [`sleap-track`](https://sleap.ai/develop/guides/cli.html#sleap-track). The nodes that switch around likely have a low confidence score. We can filter out nodes with low scores by increasing the value of `--peak-threshold` (default is 0.2). Linking a similar discussion [here](https://github.com/talmolab/sleap/discussions/991#discussioncomment-3852754).

You can then interpolate missing nodes in the analysis phase if needed.

Let me know if this helps!

Thanks,
Liezl

**Comment by auesro** [comment]
Hi @roomrys,

That sounds like a great option. I will follow the advice.
Do we have access to that parameter in the GUI or a python notebook? Or is it restricted to the CLI?

Thanks!

**Reply by roomrys** [reply]
Hi @auesro,

Unfortunately, the parameter is not exposed in the GUI. Would you like to see it in the GUI? 

You can call the [`sleap.nn.inference.main()`](https://sleap.ai/api/sleap.nn.inference.html#sleap.nn.inference.main) function which takes in a list of `sleap-track` arguments or call `!sleap-track <your-arguments-here>` in a python script/notebook (the `!` tells python to interpret that line as a CLI command).

Let me know if this helps!

Thanks,
Liezl

**Reply by auesro** [reply]
Hi @roomrys 

Thanks for the clarification. Just wondering how to best proceed efficiently. That solves it!

Cheers!

A

**Comment by auesro** [comment]
Dear @roomrys 

We have been playing with the `--peak_threshold` parameter a bit. What would be the best approach in order to get as clean and realistic tracking data as possible (considering only 1 animal in the video)?:
-Lower `peak_threshold` to include as many body parts as possible-> clean incomplete instances through tracking parameters/proofreading->interpolate missing parts.
-Increase `peak_threshold` to include only "trusted" body parts->tracking/proofreading will not have much impact in the dataset-> interpolate missing parts.

And finally...any good method/algorithm for the interpolation?

Thanks!

A

**Reply by sean-afshar** [reply]
Hi, I am one of Liezl's coworkers and I use SLEAP quite a bit myself. She asked me to give my input on this problem because I have had similar issues. 

I usually increase the threshold and interpolate what I need. You can try this notebook, https://sleap.ai/notebooks/Analysis_examples.html, the code block under the fill missing values section should show how to use basic linear interpolation. You can also try using a Kalman filter to predict future points, which I believe is accessible via SciPy.

**Reply by auesro** [reply]
Hi @sean-afshar

Thanks for your input, we will proceed then by increasing the peak value and interpolate from there.

Cheers,

A
